packages/docusaurus-theme-common/src/utils/storageUtils.ts

```diff
@@ -5,12 +5,57 @@
  * LICENSE file in the root directory of this source tree.
  */
 
+import {useCallback, useRef} from 'react';
+// @ts-expect-error: TODO temp error until React 18 upgrade
+import {useSyncExternalStore} from 'use-sync-external-store/shim';
+
 const StorageTypes = ['localStorage', 'sessionStorage', 'none'] as const;
 
 export type StorageType = typeof StorageTypes[number];
 
 const DefaultStorageType: StorageType = 'localStorage';
 
+// Because we need our own event system to allow subscribing to events
+// window.addEventListener('storage') only works for different windows...
+// see https://stackoverflow.com/questions/26974084/listen-for-changes-with-localstorage-on-the-same-window
+interface StorageSlotChangeEvent {
+  key: string;
+  value: string | null;
+  storage: Storage;
+}
+
+const CustomEventName = 'StorageSlotChangeEvent';
+
+function dispatchChangeEvent({
+  key,
+  value,
+  storage,
+}: {
+  key: string;
+  value: string | null;
+  storage: Storage;
+}) {
+  const event = new CustomEvent<StorageSlotChangeEvent>(CustomEventName, {
+    detail: {
+      key,
+      value,
+      storage,
+    },
+  });
+  window.dispatchEvent(event);
+}
+
+function subscribeChangeEvents(
+  onChange: (event: StorageSlotChangeEvent) => void,
+): () => void {
+  const listener = (domEvent: CustomEvent<StorageSlotChangeEvent>) =>
+    onChange(domEvent.detail);
+  // @ts-expect-error: TODO custom events are annoying to type
+  window.addEventListener(CustomEventName, listener);
+  // @ts-expect-error: TODO custom events are annoying to type
+  return () => window.removeEventListener(CustomEventName, listener);
```

Comment:
#### Suggestion
Maybe we can create a new event emitter instead of using `window`?
It might be easier to type.
https://basarat.gitbook.io/typescript/main-1/typed-event

Reply:
👍 

Actually I hesitated to bring my own tiny eventbus lib:
https://github.com/slorber/micro-typed-events

Another approach could be to emit regular storage events fired manually for the same window:
https://stackoverflow.com/a/71177640/82609
The interested is that you can use `addEventListener("storage")` and it would work for "same windows" localstorage changes, but also "cross window" localstorage changes, so it's an extra useful feature to have (groupId sync between different tabs)